### PR TITLE
Dynamical_matrix and third_order features+update

### PR DESCRIPTION
**Pull request by charlessievers**
**Summary**

Further optimize third order, add folding to both commands, add output cutoff control for third order, add GPU support to both commands, reformat dynamical_matrix command to have 2 space indentation, and add force constant matrix option to dynamical_matrix command.

**Author(s)**

Charlie Sievers, UC Davis, charliesievers@cox.net

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.

**Further Information, Files, and Links**

Feature is not complete yet. 

Folding is discussed here: https://dannyvanpoucke.be/folding-phonons/
The fold options considers the interactions of the supercell (N atoms) w.r.t displacements of the original cell (n atoms). This produces a dynamical matrix of size (3nx3N). The original cell is designated by a user defined group, and the supercell is always considered to be the group all.

### CODE REVIEW

File: src/USER-PHONON/third_order.cpp
Comment (charlessievers) [inline src/USER-PHONON/third_order.cpp:282]: This might be a tall order, but I doubt it. I would like to check if atom j is a neighbor of atom i. It is needless to continue in the for loop otherwise and this addition would heavily reduce the number of force compute calls.

I am currently trying to reading through compute coord/atom to see how the compute forms its own neighbor list, thanks to a great suggestion by @athomps. Any additional advice is welcome.

Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:282]: @charlessievers i would rather look at commands like `create_bonds` or `delete_atoms` for how to request and build a neighborlist from inside a command. This can be a full or a half neighbor list.

Short of using that, there also is the option to do spatial binning aka cell lists (if for some reason neighbor lists are not practical). You can globally bin all atoms (an O(N) process) and then only have to compute the closest distance between two bins to decide whether all pairs of atoms formed between the two bins are outside of a given cutoff and can be skipped. that would be a significant savings although not as much as using neighbor lists.

Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp:282]: I will look at those, thank you!

Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp:282]: Bear with me on my next question. Does ilist always contain every owned local non-ghost id?

Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:282]: > Bear with me on my next question. Does ilist always contain every owned local non-ghost id?

most of the time. not when atoms are explicitly excluded from the neighbor list. you can easily check for that with comparing `inum` to `atom->nlocal`

Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp:282]: Excellent, first time really looking into LAMMPS neighbor lists so I am sorry for the basic questions.

Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:282]: also ilist does not contain ids (aka tags) but local indices, i.e. 0 to atom->nlocal. if you want the id you need `atom->tag[ilist[j] & NEIGHMASK]`. also you need to watch out for special_bond bits, hence the application of NEIGHMASK.

Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp:282]: I assume, based on what you just said, there is no trivial atom->map() equivalent to go from tag to ilist index "ii" or jlist index "jj"

Comment (martok) [reply src/USER-PHONON/third_order.cpp:282]: I thought about this a while ago, but somehow got stuck on the neighbors-of-neighbors problem: moving one atom can change the force on an atom two neighbors over by chaning \rho of the inbetween one. I think only atoms at 2* force cutoff can be safely ignored?

Comment (athomps) [reply src/USER-PHONON/third_order.cpp:282]: That is correct. The necessary but not sufficient condition for non-zero dFj/dRi is:
    there exists atom k s.t. i and j are neighbors of k, where k is always a neighbor of itself.
A fairly straightforward way to find all non-zero dFj/dRi is to first build a full neighbor list, then loop over atoms k, then do a double loop over neighbors of k, adding to a new list storing all ordered (i,j) pairs that satisfy the above criterion.

File: src/USER-PHONON/dynamical_matrix.cpp
Comment (charlessievers) [reply src/USER-PHONON/dynamical_matrix.cpp:275]: Are the new calls "%ill"?

**Comment by charlessievers** [comment]
> The files are already diverging in minor formatting things and IMO, that's a maintainance hazard. But Axel will have to call the shot on that...

If there is a clean way to do this, I am all ears. Thank you for reviewing the PR by the way, I appreciate it.

File: src/USER-PHONON/third_order.cpp
Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:282]: > I assume, based on what you just said, there is no trivial atom->map() equivalent to go from tag to ilist index "ii".

why would you need that? is there anything that forces you in looping over tags in order? you can always get the tag from atom->tag[i] or atom->tag[j].

Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:282]: > I thought about this a while ago, but somehow got stuck on the neighbors-of-neighbors problem: moving one atom can change the force on an atom two neighbors over by chaning \rho of the inbetween one. I think only atoms at 2* force cutoff can be safely ignored?

unless you have long-range electrostatics or a model with dihedrals and a short pairwise cutoff (in which case you should get a warning to increase the communication cutoff).

**Comment by akohlmey** [comment]
> The files are already diverging in minor formatting things and IMO, that's a maintainance hazard. But Axel will have to call the shot on that...

with contributed code, the primary choice is on the person contributing the code unless it is something that the LAMMPS core developers take an active interest in maintaining. In general, we favor anything that reduces code duplication for as long as it does not result in significant additional complexity. the fix_nh.cpp file is an example where the complexity and the code reuse cancelling each other. so we have started to recommend people with additional features to create copies with all optional and unused parts removed.

File: src/USER-PHONON/third_order.cpp
Comment (akohlmey) [reply src/USER-PHONON/third_order.cpp:271]: and possibly also `utils::logmesg()` to have the text printed to both screen and logfile at no extra coding effort.

**Comment by akohlmey** [comment]
> > The files are already diverging in minor formatting things and IMO, that's a maintainance hazard. But Axel will have to call the shot on that...
> 
> If there is a clean way to do this, I am all ears. Thank you for reviewing the PR by the way, I appreciate it.

Typically one would create a base class with all shared functionality (and without using the CommandStyle() macro) , and then implement each command as a derived class. you may need to rename functions that are shared or break them down to pieces.

There are plenty of examples in LAMMPS. What comes to mind right now are KSpace and PPPM.

File: src/USER-PHONON/third_order.cpp
Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp]: Just pushed a **draft** of my solution. I only search through first neighbors. I realize that I need to loop over their neighbors as well, but I figured it would be nice to have a draft to work off of.

Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp]: New draft produced a neighbor/neighbor_of_neighbor list (extended neighbor list). This list is indexed by atom tag and yields a pointer to an array that contains the atom tags of the extended neighbors. Next step for me is to make this depend on the third order's group map member.

File: src/USER-PHONON/third_order.cpp
Comment (charlessievers) [reply src/USER-PHONON/third_order.cpp:299]: fmtlib looks a lot like python formatting, I like it already.

**Comment by charlessievers** [comment]
> Typically one would create a base class with all shared functionality (and without using the CommandStyle() macro) , and then implement each command as a derived class. you may need to rename functions that are shared or break them down to pieces.
> 
> There are plenty of examples in LAMMPS. What comes to mind right now are KSpace and PPPM.

Thanks @akohlmey, I will look into these. I think combining these could be pretty clean.

I was thinking of adding GPU support. I believe all I need to do is to transfer the forces from device to host, any advice on what file to look at?

**Comment by akohlmey** [comment]
> I was thinking of adding GPU support. I believe all I need to do is to transfer the forces from device to host, any advice on what file to look at?

supporting the GPU package (or USER-INTEL) should be straightforward and not needing much work, if any. KOKKOS is a different story, for that you probably need to implement a derived class, so that you can optimize data handling (it may work without, but be less efficient). It would probably best to look at the minimizer classes or the verlet classes for hints of what is needed in addition to plain styles.

**Comment by charlessievers** [comment]
> supporting the GPU package (or USER-INTEL) should be straightforward and not needing much work, if any. KOKKOS is a different story, for that you probably need to implement a derived class, so that you can optimize data handling (it may work without, but be less efficient). It would probably best to look at the minimizer classes or the verlet classes for hints of what is needed in addition to plain styles.

I see that I needed to setup both fix_omp and fix_gpu. I now have omp working.
Does the GPU package affect run at any other times besides compute and post_force?

**Comment by akohlmey** [comment]
@charlessievers can you give us an update on the status of this PR? Thanks.

**Comment by charlessievers** [comment]
Hello @akohlmey, I am still working on this. I actually just picked it back up this week oddly enough.

The current status:
1. Added cell folding (dynamical matrix of a unit cell in a supercell)
2. Greatly reduced time to calculate the third order
3. Added OMP support
4. Working on GPU support
5. KOKKOS support is up next
6. Finally, I will reorganize dynamical matrix and third order to be more maintainable.

**Comment by charlessievers** [comment]
@akohlmey and @wmbrownIntel would either of you point me to where intel style potentials unload force data into atom->f

**Comment by akohlmey** [comment]
> @akohlmey and @wmbrownIntel would either of you point me to where intel style potentials unload force data into atom->f

I don't know much about the INTEL package and it is not the most readable code in LAMMPS, too. As far as I recall, the synchronization of the force arrays is done by fix intel. The `FixIntel::add_results()` function seems to be where it starts.

**Comment by charlessievers** [comment]
@akohlmey Thanks! Looks like a great place for me to start, I think based on that I need some additional modify methods in my force update method. @wmbrownIntel Any advice on which modify methods I absolutely need?

**Comment by charlessievers** [comment]
@akohlmey I think I have two problems, one is fixed by adding modify->pre_reverse, but the other I believe might have to do with clearing out forces. Do you see anything in DynamicalMatrix::force_clear() that is clearly lacking when it comes to intel potentials? If not, I will keep looking.

**Comment by akohlmey** [comment]
> @akohlmey I think I have two problems, one is fixed by adding modify->pre_reverse, but the other I believe might have to do with clearing out forces. Do you see anything in DynamicalMatrix::force_clear() that is clearly lacking when it comes to intel potentials? If not, I will keep looking.

It looks to me that your handling of the `external_force_clear` flag is inconsistent. You have it set to 1 in the constructor (where it probably should be set to 0) and then unconditionally set it to 0 in setup() but "late" (although possibly not too late).
If you look at minimizers or integrators. It would be set to 0 unless fix omp is defined. That won't affect the INTEL package, though. It looks as if those styles only need the pre-reverse processing.

**Comment by charlessievers** [comment]
@akohlmey That is a good point I'll change that to keep things consistent with min.cpp and verlet.cpp. 

In my tests, I am seeing numerical values in atom->f after pre_reverse and before the reverse comm, however, after reverse comm the values in atom->f are zero. So I know that I fix_intel is running, computing forces, and offloading forces onto atom->f. Maybe I have something wrong in my setup that is preventing fix_intel from calculating the forces properly?

**Comment by charlessievers** [comment]
@sjplimp As I was adding in modify->pre_reverse to the dynamical matrix command, I was wondering if fix numdiff needed this as well. So I tested it and I am pretty sure it needs to be added in there as well.

**Comment by sjplimp** [comment]
@charlessievers make sense, you're adding that to this PR ?

**Comment by charlessievers** [comment]
> @charlessievers make sense, you're adding that to this PR ?

Sure can, if that is alright with @akohlmey. If not, I will make a separate PR.

@akohlmey I found the Intel potential issue, see line 428 in dynamical_matrix.cpp. I still need to update third_order accordingly. Thanks for the help as always.

**Comment by akohlmey** [comment]
> Sure can, if that is alright with @akohlmey. If not, I will make a separate PR.

Depends on how long it will take until this PR is ready to be merged. If soon, then I won't mind to have those combined. Just add a reminder about it at the top.

File: src/PHONON/dynamical_matrix.cpp
Comment (akohlmey) [inline src/PHONON/dynamical_matrix.cpp:90]: @charlessievers please note that as of PR #3001 the use of `Modify::find_fix()` is deprecated in favor of `Modify::get_fix_by_id()` (it will keep working for a while, but I have started to phase its use out). This makes the naming more consistent and saves one line as you can do:
```
if (!modify->get_fix_by_id("package_omp")) external_force_clear = 1;
```

Comment (charlessievers) [reply src/PHONON/dynamical_matrix.cpp:90]: I will merge the develop branch and change those lines accordingly then, thanks for letting me know.

**Comment by charlessievers** [comment]
Hey, @stanmoore1 I was wondering if you would give me a bit of advice on allowing kokkos potentials to work with the dynamical matrix command. I understand that I want to offload force from the compute device to the host device.

**Comment by stanmoore1** [comment]
> Hey, @stanmoore1 I was wondering if you would give me a bit of advice on allowing kokkos potentials to work with the dynamical matrix command. I understand that I want to offload force from the compute device to the host device.

@charlessievers sure, to get a minimum working case you probably just need to add a `sync-to-host` command before your computations and a `modified-on-host` command afterwards. We can talk more offline if you want.

**Comment by charlessievers** [comment]
@stanmoore1 I had on error on my end. The kokkos dynamical matrix command works so far, I will need to do a bit more testing.

**Review by stanmoore1 — APPROVED** [review]

File: src/KOKKOS/dynamical_matrix_kokkos.h
Comment (stanmoore1) [inline src/KOKKOS/dynamical_matrix_kokkos.h:1]: Needs the LAMMPS header.

**Review by stanmoore1 — CHANGES_REQUESTED** [review]

**Comment by charlessievers** [comment]
@ndtrung81 I am getting an error in my dynamical matrix command that I am not getting when I run the example in.tersoff in examples/tersoff.

```mpirun -n 1 lmp -sf gpu -pk gpu 1 -in examples/PACKAGES/phonon/dynamical_matrix_command/Silicon/in.silicon```
gives
```
No protocol specified
LAMMPS (27 Oct 2021)
OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (src/comm.cpp:98)
  using 1 OpenMP thread(s) per MPI task
Reading data file ...
  orthogonal box = (0 0 0) to (5.431 5.431 5.431)
  1 by 1 by 1 MPI processor grid
  reading atoms ...
  8 atoms
Finding 1-2 1-3 1-4 neighbors ...
  special bond factors lj:    0        0        0       
  special bond factors coul:  0        0        0       
     0 = max # of 1-2 neighbors
     0 = max # of 1-3 neighbors
     0 = max # of 1-4 neighbors
     1 = max # of special neighbors
  special bonds CPU = 0.000 seconds
  read_data CPU = 0.001 seconds
Reading tersoff potential file SiCGe.tersoff with DATE: 2009-03-18

--------------------------------------------------------------------------
- Using acceleration for tersoff/gpu:
-  with 1 proc(s) per device.
-  Horizontal vector operations: ENABLED
-  Shared memory system: No
--------------------------------------------------------------------------
Device 0: NVIDIA GeForce RTX 2080 Ti, 68 CUs, 11/11 GB, 1.5 GHZ (Double Precision)
--------------------------------------------------------------------------

Initializing Device and compiling on process 0...Done.
Initializing Device 0 on core 0...Done.

WARNING: Increasing communication cutoff for GPU style (src/GPU/pair_tersoff_gpu.cpp:251)
Cuda driver error 700 in call at file '/home/csievers/lammps/lib/gpu/geryon/nvd_timer.h' in line 76.
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode -1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```
Any idea what could be going wrong? No worries if not.

**Comment by charlessievers** [comment]
@ndtrung81 sorry if I wasted any of your time, I figured it out. I need to initialize the device similar to fix_gpu or one can just put `run 0` before using the command.

@akohlmey, I found that if I add `neighbor->decide();` at the top of `update_force()` intel potentials will work, but gpu potentials will output forces = 0. Do you have any advice?

**Comment by akohlmey** [comment]
> @akohlmey, I found that if I add `neighbor->decide();` at the top of `update_force()` intel potentials will work, but gpu potentials will output forces = 0. Do you have any advice?

No. I am not very familiar with the calling sequence details of the fix and requirements of either the GPU or the INTEL code. This is a question for @wmbrownIntel and @ndtrung81

**Comment by ndtrung81** [comment]
@charlessievers sorry I couldn't reproduce the error you got when running examples/PACKAGES/phonon/dynamical_matrix_command/Silicon/in.silicon.
with the recent commit on the master branch (SHA1 ID: b3fcda3214e16c35d9e3aa2d7fa4dd7e8b47d748). I built LAMMPS with CMake, CUDA 11.4, double precision, sm_60.

For the GPU package, the device is initialized within the FixGPU constructor, which is invoked by the package gpu command, or just by using -sf gpu at the command line. I don't think you need anything else (like adding run 0).

Also, you wrote that "run the example in.tersoff in examples/tersoff", but the actual run command line is for examples/PACKAGES/phonon/dynamical_matrix_command/Silicon/in.silicon.

**Comment by charlessievers** [comment]
Hello @ndtrung81,

Sorry for your time, I had an error with this PR not the master branch. 
I was comparing the two examples to one another since one worked and one did not and they both used tersoff potentials.

I added `neighbor->decide();` to `force_update` (in this PR) for intel potentials to work and that directly affects GPU forces, any idea why that might be the case?

**Comment by ndtrung81** [comment]
@charlessievers can you try using "-pk gpu 1 neigh no" to see if the error persists? neighbor->decide() changes the ago value which is used with the default neigh yes to build neighbor list or not.

**Comment by charlessievers** [comment]
@ndtrung81 The error, which is producing all forces = 0, still persists.

**Comment by ndtrung81** [comment]
@charlessievers the reason for zero forces with the gpu pair styles in this case is that their neighbor list is neither rebuilt, nor reset as neighbor->ago becomes nonzero after the call to neighbor->decide(). 

Maybe you could try something like this in update_force()

    #ifdef LMP_INTEL
    neighbor->decide();
    #endif

or

    neighbor->decide();
    neighbor->ago = 0;

**Comment by akohlmey** [comment]
> @charlessievers the reason for zero forces with the gpu pair styles in this case is that their neighbor list is neither rebuilt, nor reset as neighbor->ago becomes nonzero after the call to neighbor->decide().
> 
> Maybe you could try something like this in update_force()
> 
> ```
> #ifdef LMP_INTEL
> neighbor->decide();
> #endif
> ```
> 
> or
> 
> ```
> neighbor->decide();
> neighbor->ago = 0;
> ```

Should the need to reset `neighbor->ago` to zero depend on whether or not the GPU package is used, then this has to be a conditional(!) run-time choice, e.g. which can be detected with something like `(modify->get_fix_by_id("package_gpu")) ? true : false`.

**Comment by ndtrung81** [comment]
@akohlmey right, neighbor->ago needs to be reset to zero only when the GPU package is in use. What you suggest for a run-time check makes sense to me.

**Comment by charlessievers** [comment]
@ndtrung81 @akohlmey Sorry, I am just getting back to this. I believe in all cases except intel, neighbor->decide() would be extraneous. Update->force() is a frequently called method so I would prefer to reduce extraneous calls. I will most likely invert the if statement to look like the expression below:
```
  neighbor->ago = 0;
  if ((modify->get_fix_by_id("package_intel")) ? true : false)
    neighbor->decide();
```

**Comment by charlessievers** [comment]
@akohlmey I got all the bugs that I know of fixed and all the accelerator packages I want working with my command. I think it is finally ready to be considered for a merge.

**Comment by akohlmey** [comment]
@stanmoore1 @rbberger this PR is currently blocked from merging because of your requests for changes. I think those have been satisfied by now, so please check and - if yes - approve.

**Review by rbberger — APPROVED** [review]

**Review by athomps — APPROVED** [review]
I approve

**Review by stanmoore1 — APPROVED** [review]

**Pull request merged**
